L1Trigger/TrackTrigger/plugins/TTStubBuilder.cc

```diff
@@ -148,65 +172,172 @@ void TTStubBuilder< Ref_Phase2TrackerDigi_ >::produce( edm::Event& iEvent, const
         }
         else
         {
+	  bool FEreject = false;
           /// This means that only some of them do
           /// Put in the temporary output
-          int chip = tempTTStub.getTriggerPosition() / chipSize; /// Find out which ASIC
-          if ( moduleStubs.find( chip ) == moduleStubs.end() ) /// Already a stub for this ASIC?
+          MeasurementPoint mp0 = tempTTStub.getClusterRef(0)->findAverageLocalCoordinates();
+          int seg       = static_cast<int>(mp0.y());
+	  if (isPS) seg = seg/16;
+	  int chip      = 1000*nmod+10*int(tempTTStub.getTriggerPosition()/chipSize)+seg; /// Find out which MPA/CBC ASIC
+          int CIC_chip  = 10*nmod+seg; /// Find out which CIC ASIC
+
+	  // First look is the stub is passing trough the very front end (CBC/MPA)
+	  (isPS)
+	    ? maxStubs = maxStubs_PS
+	    : maxStubs = maxStubs_2S;
+
+	  if (isPS) // MPA
+	  {
+	    if ( moduleStubs_MPA.find( chip ) == moduleStubs_MPA.end() ) /// Already a stub for this ASIC?
+	    {
+	      /// No, so new entry
+	      moduleStubs_MPA.insert( std::pair< int, int >( chip, 1 ) );
+	    }
+	    else
+	    {
+	      if ( moduleStubs_MPA[chip] < int(maxStubs) )
+	      {
+		++moduleStubs_MPA[chip];
+	      }
+	      else
+	      {
+		TTStub< Ref_Phase2TrackerDigi_ > tempTTStub2( tempTTStub.getDetId() );
+
+		tempTTStub2.addClusterRef( (tempTTStub.getClusterRef(0)) );
+		tempTTStub2.addClusterRef( (tempTTStub.getClusterRef(1)) );
+
+		// We flag the rejected stub this way, so that the bend will appear correct.
+		tempTTStub2.setTriggerDisplacement( 500+2.*tempTTStub.getTriggerDisplacement() );
+		tempTTStub2.setTriggerOffset( 500+2.*tempTTStub.getTriggerOffset() ); 
+		tempTTStub2.setRealTriggerOffset( 500+2.*tempTTStub.getRealTriggerOffset() );
+		tempTTStub2.setHardwareBend( tempTTStub.getHardwareBend() );
+
+		tempInner.push_back( *(tempTTStub2.getClusterRef(0)) );
+		tempOuter.push_back( *(tempTTStub2.getClusterRef(1)) );
+		tempAccepted.push_back( tempTTStub2 );
+		FEreject = true;
+	      }
+	    }
+	  }
+	  else // CBC
+	  {
+	    if ( moduleStubs_CBC.find( chip ) == moduleStubs_CBC.end() ) /// Already a stub for this ASIC?
```

Comment:
make this logic into a separate function to avoid complete code duplication for MPA and CBC cases

Reply:
Partly done, redundancy was reduced but no new function created.